client/src/services/link.js

```diff
@@ -44,104 +43,89 @@ export function LinkProvider(props) {
-          const res = await getPublicToken(itemId);
-          token = res.data.public_token;
-        } else {
-          hasRequested.current.byUser[userId] = true;
-        }
-
-        const handler = await window.Plaid.create({
-          apiVersion: 'v2',
-          clientName: 'Pattern',
-          env: PLAID_ENV,
-          key: process.env.REACT_APP_PLAID_PUBLIC_KEY,
-          onEvent: logEvent,
-          onLoad: () => {
-            logEvent('onLoad', {});
-            if (isUpdate) {
-              dispatch([types.LINK_UPDATE_MODE_LOADED, { id: itemId }]);
-            } else {
-              dispatch([types.LINK_LOADED, { id: userId }]);
-            }
-          },
-          onSuccess: async (
-            publicToken,
-            { institution, accounts, link_session_id }
-          ) => {
-            logEvent('onSuccess', { institution, accounts, link_session_id });
-            await postLinkEvent({
-              userId,
-              link_session_id,
-              type: 'success',
-            });
-            if (isUpdate) {
-              await setItemState(itemId, 'good');
-              getItemById(itemId, true);
-            } else {
-              await exchangeToken(publicToken, institution, accounts, userId);
-              getItemsByUser(userId, true);
-            }
-
-            if (window.location.pathname === '/') {
-              window.location.href = `/user/${userId}/items`;
-            }
-          },
-          onExit: async (
+      const isUpdate = itemId != null;
+
+      const linkTokenRes = await getLinkToken({ itemId, userId });
+      if (isUpdate) {
+        hasRequested.current.byItem[itemId] = true;
+      } else {
+        hasRequested.current.byUser[userId] = true;
+      }
+
+      const handler = await window.Plaid.create({
+        apiVersion: 'v2',
+        clientName: 'Pattern',
+        env: PLAID_ENV,
+        token: linkTokenRes.data.link_token,
+        onEvent: logEvent,
+        onLoad: () => {
+          logEvent('onLoad', {});
+          if (isUpdate) {
+            dispatch([types.LINK_UPDATE_MODE_LOADED, {id: itemId}]);
+          } else {
+            dispatch([types.LINK_LOADED, {id: userId}]);
+          }
+        },
+        onSuccess: async (
+          publicToken,
+          {institution, accounts, link_session_id}
+        ) => {
+          logEvent('onSuccess', {institution, accounts, link_session_id});
+          await postLinkEvent({
+            userId,
+            link_session_id,
+            type: 'success',
+          });
+          if (isUpdate) {
+            await setItemState(itemId, 'good');
+            getItemById(itemId, true);
+          } else {
+            await exchangeToken(publicToken, institution, accounts, userId);
+            getItemsByUser(userId, true);
+          }
+
+          if (window.location.pathname === '/') {
+            window.location.href = `/user/${userId}/items`;
+          }
+        },
+        onExit: async (
```

Comment:
In this `onExit` handler, perhaps we should show the best practice of handling the `INVALID_LINK_TOKEN` error_code when a link_token is invalidated (too many attempts, expired, etc).

```
if (error != null && error.error_code === 'INVALID_LINK_TOKEN`) {
    // gets rid of the old iframe
    handler.destroy();
    // generate a new token and reinitialize Link
    handler.create({... oldConfigs, token: await getLinkToken({ userId, itemId }) });
}
```